src/legacy/ui/public/vis/vis_types/vislib_vis_legend/vislib_vis_legend.tsx

```diff
@@ -0,0 +1,246 @@
+
+export interface VisLegendProps {
+  vis: any;
+  vislibVis: any;
+  visData: any;
+  uiState: any;
+}
+
+export interface VisLegendState {
+  labels: any[];
+  tableAggs: any[];
+  selectedLabel: string | null;
+}
+
+export class VisLegend extends PureComponent<VisLegendProps, VisLegendState> {
+  static propTypes = {
+    vis: PropTypes.object,
+    visData: PropTypes.object,
+    uiState: PropTypes.object,
+  };
+
+  legendId = htmlIdGenerator()('legend');
+  getColor: (label: string) => string = () => '';
+
+  constructor(props: VisLegendProps) {
+    super(props);
+
+    this.state = {
+      labels: [],
+      tableAggs: [],
+      selectedLabel: null,
+    };
+  }
+
+  componentDidMount() {
+    this.refresh();
+  }
+
+  toggleLegend = () => {
+    const bwcAddLegend = this.props.vis.params.addLegend;
+    const bwcLegendStateDefault = bwcAddLegend == null ? true : bwcAddLegend;
+    const newOpen = !this.props.uiState.get('vis.legendOpen', bwcLegendStateDefault);
+    this.props.uiState.set('vis.legendOpen', newOpen);
+  };
+
+  setColor = (label: string, color: string) => (event: BaseSyntheticEvent) => {
+    if ((event as KeyboardEvent).keyCode && (event as KeyboardEvent).keyCode !== keyCodes.ENTER) {
+      return;
+    }
+
+    const colors = this.props.uiState.get('vis.colors') || {};
+    if (colors[label] === color) delete colors[label];
+    else colors[label] = color;
+    this.props.uiState.setSilent('vis.colors', null);
+    this.props.uiState.set('vis.colors', colors);
+    this.props.uiState.emit('colorChanged');
+    this.refresh();
+  };
+
+  filter = ({ values: data }: LegendItem, negate: boolean) => () => {
+    this.props.vis.API.events.filter({ data, negate });
+  };
+
+  canFilter = (item: LegendItem): boolean => {
+    if (CUSTOM_LEGEND_VIS_TYPES.includes(this.props.vislibVis.visConfigArgs.type)) {
+      return false;
+    }
+    const filters = createFiltersFromEvent({ aggConfigs: this.state.tableAggs, data: item.values });
+    return Boolean(filters.length);
+  };
+
+  toggleDetails = (label: string | null) => (event?: BaseSyntheticEvent) => {
+    if (
+      event &&
+      (event as KeyboardEvent).keyCode &&
+      (event as KeyboardEvent).keyCode !== keyCodes.ENTER
+    ) {
+      return;
+    }
+    this.setState({ selectedLabel: this.state.selectedLabel === label ? null : label });
+  };
+
+  getSeriesLabels = (data: any[]) => {
+    const values = data.map(chart => chart.series).reduce((a, b) => a.concat(b), []);
+
+    return compact(uniq(values, 'label')).map((label: any) => ({
+      ...label,
+      values: [label.values[0].seriesRaw],
+    }));
+  };
+
+  // Most of these functions were moved directly from the old Legend class. Not a fan of this.
+  getLabels = (data: any, type: string) => {
+    if (!data) return [];
+    data = data.columns || data.rows || [data];
+
+    if (type === 'pie') return Data.prototype.pieNames(data);
```

Comment:
nit: It might be worth adding a method to the Data class that will return the array accessed here so that we don't need to call anything off of a classes prototype in a `.tsx` file.

Reply:
It seems like this is calling this class method as if it were a static method, but it uses an instance method inside the `pieNames` method.

https://github.com/elastic/kibana/blob/7fddc234ae51ba9acaee1f3f2c477d560df55efc/src/legacy/ui/public/vislib/lib/data.js#L399-L418

I didn't change this to avoid unknown issues but if you see an easy way to do it, lmk.